src/manifest.rs

```diff
@@ -144,5 +156,119 @@ pub struct DependencyManifest {
     /// Version requirement in the buffrs format, currently only supports pinning
     pub version: VersionReq,
     /// Artifactory repository to pull dependency from
-    pub repository: String,
+    pub repository: Repository,
+    /// Artifactory registry to pull from
+    pub registry: RegistryUri,
+}
+
+#[derive(Debug, Clone, Hash, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RegistryUri(Url);
+
+impl std::ops::Deref for RegistryUri {
+    type Target = Url;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for RegistryUri {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl Display for RegistryUri {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl FromStr for RegistryUri {
+    type Err = eyre::Error;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        let slf = Self(Url::from_str(value)?);
+        slf.sanity_check_url()?;
+        Ok(slf)
+    }
+}
+
+impl RegistryUri {
+    pub fn sanity_check_url(&self) -> eyre::Result<()> {
+        tracing::debug!(
+            "checking that url begins with http or https: {}",
+            self.0.scheme()
+        );
+        ensure!(
+            self.0.scheme() == "http" || self.0.scheme() == "https",
+            "The self.0 must start with http:// or https://"
+        );
+
+        tracing::debug!(
+            "checking that self.0 ends with /artifactory: {}",
+            self.0.path()
+        );
+        ensure!(
+            self.0.path().ends_with("/artifactory"),
+            "The url must end with '/artifactory'"
+        );
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Default, Clone, Hash, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RegistryToken(pub String);
+
+impl std::ops::Deref for RegistryToken {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for RegistryToken {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl FromStr for RegistryToken {
+    type Err = eyre::Error;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        Ok(Self(value.to_owned()))
+    }
+}
+
+#[derive(Debug, Clone, Hash, Serialize, Deserialize, PartialEq, Eq)]
+pub struct Repository(pub String);
+
+impl std::ops::Deref for Repository {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Display for Repository {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<&str> for Repository {
+    fn from(value: &str) -> Self {
+        Self(value.into())
+    }
+}
+
+/// Authentication data and settings for an artifactory registry
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
```

Comment:
```suggestion
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
```

credentials should never be printed – if you want them to be printed you should impl debug manually / omitting the token